blockchain_wasi.md

```diff
@@ -0,0 +1,200 @@
+  discarded.
+
+  On-chain storage MUST be exposed as a virtual file system.
+
+  A specific namespace SHOULD be reserved within the virtual file system
+  for exposing blockchain-specific functionality.
+  This functionality SHOULD be identified as `/opt/CHAIN`, where
+  `CHAIN` MAY be either a common short name identifying the network
+  OR the network identifier adopted by the chain.
+
+  * `/opt/CHAIN/ADDRESS` SHOULD should be used as the conventional
+  location for references to other contracts. Within these directories
+  the following files SHOULD be exposed:
+  * `code` - the code at that address, if any.
+  * `sock` - A file descriptor for cross-contract calls to that address.
+  * `balance` - The balance at that address.
+
+  The context of the execution is most naturally exposed through the WASI
+  `environ_get` call. This includes context of the exeuction in
+  a blockchain context, including the gas constraints, the
+  initiator of execution, and metadata of the transaction
+  initiating the execution. Such metadata SHOULD be exposed as
+  environmental variables.
+
+5. Standards Considerations
+
+5.1 File System Nodes
+
+  Blockchain-specific functionality is exposed in the `/opt`
+  namespace where it does not conflict with existing file
+  system semantics.
+
+  The home directory MUST be writable, and is mapped to persistant
+  storage that will be committed with a transaction and
+  subsequently available to future transactions.
+
+  The `/tmp` directory SHOULD be used to provide a namespace that is
+  writable but ephemeral, and will not be saved between
+  transactions.
+
+  The remainder of the file system presented SHOULD be read-only,
+  and any attempts at writes SHOULD fail.
+
+  The virtual file system translates paths to addresses in
+  the underlying key-value store. In WASI, the paths and byte
+  offsets are supplied to the Wasm runtime via `__wasi_fd_open`,
+  `__wasi_fd_seek`. The runtime associates the key with the file
+  descriptor so that it can service future reads and writes
+  received from `__wasi_fd_(read|write)`.
+
+5.1.1 Deviations from Expected File System Behavior
+
+  Directory listing of the `/opt/CHAIN` directory will list 0
+  items, since the full namespace cannot meaningfuly be represented.
+  Attempts to `fd_fdstat_set_rights` (change permissions), or
+  `_filestat_set_times` (changing metadata), on blockchain-namespace
+  files will be a failure that aborts execution for safety.
+
+5.2 Socket Address and Interoperability
+
+  Opening of a socket to another blockchain executable via a
+  FD obtained from `/opt/CHAIN/ADDRESS/sock` allows for cross
+  contract calls. Data sent over the socket via `sock_send` will
+  be used as the input available on the `stdin` descriptor available
+  to the other execution.
+  The first `sock_recv` call will begin execution and block until
+  output is available.
+
+5.3 Environment
+
+  We expose the environmental variables
```

Comment:
Should also expose a variable expressing the `version` of the runtime.

Reply:
Not sure about that. The wasm module should request a particular version if there's versioning to be done, but having this available at runtime is just asking for trouble (c.f. python `six`)

Reply:
if semantics change in any way on version upgrades, this provides an escape hatch by which old contracts can shut down or force approval from an owner before continuing to allow txns

Reply:
Then why not expose an import that specifies a runtime semver? if it's not compatible, the program just won't run. If the author wants to really lock things down, she can set the semver to equality and then redeploy+migrate when a new runtime version is released.

Reply:
I don't think i care if it's an import or an env var, but we should specify how it's accessed and versioned as part of the RFC

Reply:
Okay, how about
```wat
(module
  (memory (;0;) 2)
  (global (;0;) i32 (i32.const 1024))
  (export "memory" (memory 0))
  (export "required_runtime_version" (global 0))
  (data (;0;) (i32.const 1024) "~1.0\00"))
```
where `required_runtime_version` is an export that can be picked up by the runtime, read from the memory, and parsed as a semver?

Reply:
why not the inverse? version should be able to get into the contract in some way - especially since a static export you propose would be very different for a contract to update dynamically from storage.

Reply:
I don't think that the contract should know the version since that leaks implementation details. Ideally the contract would be self contained and depend on nothing other than entropy and storage. Either the runtime supports the contract or it doesn't. We should aim for the API to be as simple as possible.

Reply:
no api is static, and to not design the ability to modify it subsequently will cause it to fail, or other non-ideal things like block numbers to be used as indications of when API versions change.

We should directly expose api version, so that there's functionality by which contacts can track if they do/don't work.

Reply:
>  functionality by which contacts can track if they do/don't work

I humbly disagree. A contract has no idea what future runtimes will look like, so there's no reasonably way besides having heavily mutable contracts to accommodate new behavior. The owner should just terminate the contract and deploy a new one.

Reply:
I think version needs to be exposed. the other APIs of this type: eth / similar chains, linux kernel sys call interface, etc. all have it, and it's valuable.